File: share/example/c++/example_algorithm.cpp
#include "aubo_sdk/rpc.h"
#ifdef WIN32
#include <Windows.h>
#endif

using namespace arcs::common_interface;
using namespace arcs::aubo_sdk;

// Template function: print variable of type std::vector<T>
template <typename T> void printVec(std::vector<T> param, std::string name) {
  std::cout << name << ": ";

  for (int i = 0; i < param.size(); i++) {
    std::cout << param.at(i);
    if (i != param.size() - 1) {
      std::cout << ", ";
    }
  }
  std::cout << std::endl;
}

// Template function: print variable of type std::vector<std::vector<T>>
template <typename T> void print2Vec(std::vector<T> param, std::string name) {
  std::cout << name << ": " << std::endl;
  for (size_t i = 0; i < param.size(); i++) {
    std::cout << "  Group " << i + 1 << ": ";
    for (size_t j = 0; j < param.at(i).size(); j++) {
      std::cout << param.at(i).at(j);

      if (j != param.at(i).size() - 1) {
        std::cout << ", ";
      } else {
        std::cout << std::endl;
      }
    }

    if ((param.size() - 1) == i) {
      std::cout << std::endl;
    }
  }
}

void exampleForwardK(RpcClientPtr impl) {
  // API call: get robot name
  auto robot_name = impl->getRobotNames().front();

  // API call: set tcp offset
  std::vector<double> offset = {0.1, 0.2, 0.3, 0.0, 0.0, 0.0};
  impl->getRobotInterface(robot_name)->getRobotConfig()->setTcpOffset(offset);

  // API call: get current joint angles
  std::vector<double> q =
      impl->getRobotInterface(robot_name)->getRobotState()->getJointPositions();

  // API call: forward kinematics to get current TCP pose
  auto result = impl->getRobotInterface(robot_name)
                    ->getRobotAlgorithm()
                    ->forwardKinematics(q);
  printVec(std::get<0>(result), "Forward kinematics: current TCP pose");
}

void exampleForwardToolK(RpcClientPtr impl) {
  // API call: get robot name
  auto robot_name = impl->getRobotNames().front();

  // API call: set tcp offset
  std::vector<double> offset = {0.1, 0.2, 0.3, 0.0, 0.0, 0.0};
  impl->getRobotInterface(robot_name)->getRobotConfig()->setTcpOffset(offset);

  // API call: get current joint angles
  std::vector<double> q =
      impl->getRobotInterface(robot_name)->getRobotState()->getJointPositions();

  // API call: forward kinematics to get flange end center pose
  auto result = impl->getRobotInterface(robot_name)
                    ->getRobotAlgorithm()
                    ->forwardToolKinematics(q);
  printVec(std::get<0>(result), "Forward kinematics: flange end center pose");
}

void exampleInverseK(RpcClientPtr impl) {
  // API call: get robot name
  auto robot_name = impl->getRobotNames().front();

  // API call: set tcp offset
  std::vector<double> offset = {0.1, 0.2, 0.3, 0.0, 0.0, 0.0};
  impl->getRobotInterface(robot_name)->getRobotConfig()->setTcpOffset(offset);

  // API call: get current joint angles
  std::vector<double> qnear =
      impl->getRobotInterface(robot_name)->getRobotState()->getJointPositions();

  // API call: get current tcp value as target pose
  std::vector<double> p =
      impl->getRobotInterface(robot_name)->getRobotState()->getTcpPose();
  printVec(p, "Current TCP pose");

  // API call: get optimal inverse kinematics solution based on TCP pose
  auto result1 = impl->getRobotInterface(robot_name)
                     ->getRobotAlgorithm()
                     ->inverseKinematics(qnear, p);
  printVec(std::get<0>(result1), "Inverse kinematics: optimal joint angles from TCP pose");

  // API call: get all inverse kinematics solutions based on TCP pose
  auto result2 = impl->getRobotInterface(robot_name)
                     ->getRobotAlgorithm()
                     ->inverseKinematicsAll(p);
  print2Vec(std::get<0>(result2), "Inverse kinematics: all joint angles from TCP pose");
}

void exampleInverseToolK(RpcClientPtr impl) {
  // API call: get robot name
  auto robot_name = impl->getRobotNames().front();

  // API call: set tcp offset
  std::vector<double> offset = {0.1, 0.2, 0.3, 0.0, 0.0, 0.0};
  impl->getRobotInterface(robot_name)->getRobotConfig()->setTcpOffset(offset);

  // API call: get current joint angles
  std::vector<double> qnear =
      impl->getRobotInterface(robot_name)->getRobotState()->getJointPositions();

  // API call: get current flange end center value as target pose
  std::vector<double> p =
      impl->getRobotInterface(robot_name)->getRobotState()->getToolPose();
  printVec(p, "Current flange end center pose");

  // API call: get optimal inverse kinematics solution based on flange end center pose
  auto result1 = impl->getRobotInterface(robot_name)
                     ->getRobotAlgorithm()
                     ->inverseToolKinematics(qnear, p);
  printVec(std::get<0>(result1),
           "Inverse kinematics: optimal joint angles from flange end center pose");

  // API call: get all inverse kinematics solutions based on flange end center pose
  auto result2 = impl->getRobotInterface(robot_name)
                     ->getRobotAlgorithm()
                     ->inverseToolKinematicsAll(p);
  print2Vec(std::get<0>(result2),
            "Inverse kinematics: all joint angles from flange end center pose");
}

#define LOCAL_IP "127.0.0.1"

int main(int argc, char **argv) {
#ifdef WIN32
  // Set Windows console output code page to UTF-8
  SetConsoleOutputCP(CP_UTF8);
#endif
  auto rpc_cli = std::make_shared<RpcClient>();
  // API call: set RPC timeout
  rpc_cli->setRequestTimeout(1000);
  // API call: connect to RPC service
  rpc_cli->connect(LOCAL_IP, 30004);
  // API call: login
  rpc_cli->login("aubo", "123456");

  // Forward kinematics: get TCP pose
  exampleForwardK(rpc_cli);

  // Forward kinematics: get flange end center pose
  exampleForwardToolK(rpc_cli);

  // Inverse kinematics: get joint angles from TCP pose
  exampleInverseK(rpc_cli);

  // Inverse kinematics: get joint angles from flange end center pose
  exampleInverseToolK(rpc_cli);

  // API call: logout
  rpc_cli->logout();
  // API call: disconnect
  rpc_cli->disconnect();

  return 0;
}

File: share/example/python/example_algorithm.py
#! /usr/bin/env python
# coding=utf-8

"""
Robot Arm Forward and Inverse Kinematics

Steps:
Step 1: Connect to the RPC service
Step 2: Log in to the robot arm
Step 3: Inverse kinematics of the robot arm
Step 4: Forward kinematics of the robot arm
"""

import pyaubo_sdk

robot_ip = "127.0.0.1"  # Server IP address
robot_port = 30004  # Port number
M_PI = 3.14159265358979323846
robot_rpc_client = pyaubo_sdk.RpcClient()
# Inverse Kinematics
def exampleInverseK(robot_name):
    robot_name = robot_rpc_client.getRobotNames()[0]  # API call: Get the robot's name
    print("Inverse Kinematics")
    # Target pose of the robot arm
    waypoint_1_p = [0.549273, -0.12094,   0.464501,
                    3.13802,  0.00122674, 1.57095]
    # Reference joint angles of the robot arm, unit: radian
    waypoint_0_q = [0.00123059, -0.26063, 1.7441,
                    0.437503,   1.56957,  0.00107586]
    # API call: Inverse kinematics
    res = robot_rpc_client.getRobotInterface(robot_name).getRobotAlgorithm().inverseKinematics(waypoint_0_q, waypoint_1_p)
    print("Inverse kinematics function return value:", res[1])
    print("Joint angles obtained from inverse kinematics:", res[0])
    pass

# Forward Kinematics
def exampleForwardK(robot_name):
    robot_name = robot_rpc_client.getRobotNames()[0]  # API call: Get the robot's name
    # Target joint angles of the robot arm
    q = [0.0012302916520035012, -0.17709153384169424, 1.3017793878477586,
         -0.08835407161244224, 1.5695657489560777, 0.0010722296812234074]
    # API call: Forward kinematics
    res = robot_rpc_client.getRobotInterface(robot_name).getRobotAlgorithm().forwardKinematics(q)
    print("Forward kinematics function return value:", res[1])
    print("Pose obtained from forward kinematics:", res[0])
    pass

    if __name__ == '__main__':
        robot_rpc_client.connect(robot_ip, robot_port)  # API call: Connect to RPC service
        if robot_rpc_client.hasConnected():
            print("Robot rpc_client connected successfully!")
            robot_rpc_client.login("aubo", "123456")  # API call: Robot arm login
            if robot_rpc_client.hasLogined():
                print("Robot rpc_client logged in successfully!")
                robot_name = robot_rpc_client.getRobotNames()[0]  # API call: Get the robot's name
                exampleInverseK(robot_name)  # Inverse kinematics
                exampleForwardK(robot_name)  # Forward kinematics

